Event based track and muon to simulation associators

Dr15Jones opened the pull request:
Replaced TrackAssociatorBase class with three classes which are obtained from the Event. This allows
modules to properly register their 'consumes'.
This change should give identical results to the previous implementation.

Dr15Jones committed:
Created a separate TrackGenAssociator

Previously, TrackAssociatorByChi2 would also match a generated particle
to a track. In order to use it code would get it from the EventSetup
via its base class and then cast it to its actual type. By creating
a seperate TrackGenAssociatorBase and TrackGenAssociatorByChi2 it
avoids the use of casting.

This is a step towards fixing the consumes problem with TrackAssociators.

Dr15Jones committed:
Created a separate MuonToSimAssociatorBase

Previously, code was getting a TrackAssociatorBase from the EventSetup
and casting it to a MuonAssociatorByHits in order to do muon association.
This change creates a separate MuonToSimAssociatorBase which provides
the correct interface. This new object can be obtained from the EventSetup
and is created by MuonAssociatorESProducer. Common code was moved to
MuonAssociatorByHitsHelper.
This change is needed as a step towards fixing the consumes problem
with TrackAssociatorBase.

Dr15Jones committed:
Made methods const correct

Made the routines used by MuonAssociatorByHitsHelper const.

Dr15Jones committed:
Split CSCHitAssociator from MuonTruth

The MuonAssociatorByHitsHelper was only using a small portion of
MuonTruth and no other user of MuonTruth used that portion. Splitting
the functionality to its own class helps reduce dependencies and avoids
unnecessary time and memory.

Dr15Jones committed:
MuonAssociatorByHitsHelper no longer knows about Event or EventSetup

Classes which use MuonAssociatorByHitsHelper are now responsible for
passing it the Event and EventSetup based resources it needs.

Dr15Jones committed:
Event based reco to sim associator classes

The various Reco to Sim assocation implementations require each
implementation to use additional information from the Event and/or
EventSetup. Inorder to accommodate that and to get the 'consumes' to
work, the association classes need to be Event data products.

This change provides the event data product for the three types of
associations and the base class for the actual implementations of
the associations. Classes which inherit from the base class are
defined in additional pacakges. This works since the associations are
transient only.

Dr15Jones committed:
Implementation of Event based Track to Sim Associations

The implementations of the associations are all copies of the
code in SimTrack/TrackAssociation where the Event or EventSetup
data products are held by the product rather than obtained from
the Event or EventSetup each call. These implement the interface
defined in SimDataFormats/Associations.

Dr15Jones committed:
Implementation of Event based Muon to Sim Associations

The implementations of the associations are all copies of the
EventSetup based product code  where the Event or EventSetup
data products are held by the product rather than obtained from
the Event or EventSetup   each call. This implement the interface
defined in SimDataFormats/Associations.

Dr15Jones committed:
Added configuration files for new Track Associator EDProducers

Dr15Jones committed:
Use reco::TrackToTrackingParticleAssociator instead of TrackAssociatorBase

The TrackAssociatorBase lives in the EventSetup but gets data from both the
edm::Event and edm::EventSetup. This is not allowed by CMSSW rules. Therefore
TrackAssociatorBase is replaced by reco::TrackToTrackingParticleAssociator which
is obtained via the edm::Event.
This solves the consumes problem associated with TrackAssociatorBase usage.

Dr15Jones committed:
Use reco::TrackToGenParticleAssociator instead of TrackGenAssociatorBase

TrackGenAssociatorBase is an EventSetup object which uses the edm::Event
to get data. This is a problem for consumes. So instead we nw use the
reco::TrackToGenParticleAssociator which does the same work but is an
edm::Event product which already contains the data it needs from the event.

Dr15Jones committed:
Use reco::MuonToTrackingParticleAssociator instead of MuonToSimAssociatorBase

MuonToSimAssociatorBase is an EventSetup object which uses the edm::Event
to get data. This is a problem for consumes. So instead    we nw use the
reco::MuonToTrackingParticleAssociator which does the same work but  is an
edm::Event product which already contains the data it needs from the event.

cmsbuild commented:
A new Pull Request was created by @Dr15Jones (Chris Jones) for CMSSW_7_4_X.

Event based track and muon to simulation associators

It involves the following packages:

Alignment/OfflineValidation
CalibTracker/SiStripCommon
Configuration/StandardSequences
DPGAnalysis/SiStripTools
FastSimulation/Validation
MuonAnalysis/MuonAssociators
RecoBTag/SoftLepton
RecoMuon/TrackingTools
RecoTracker/DebugTools
RecoVertex/ConfigurableVertexReco
RecoVertex/KalmanVertexFit
RecoVertex/KinematicFit
SLHCUpgradeSimulations/Geometry
SimDataFormats/Associations
SimMuon/MCTruth
SimTracker/TrackAssociation
SimTracker/TrackAssociatorESProducer
SimTracker/TrackAssociatorProducers
SimTracker/TrackHistory
Validation/RecoEgamma
Validation/RecoHI
Validation/RecoMuon
Validation/RecoTrack
Validation/RecoVertex


The following packages do not have a category, yet:

SimDataFormats/Associations
SimTracker/TrackAssociatorProducers

@civanch, @diguida, @lveldere, @cvuosalo, @ianna, @mdhildreth, @monttj, @cmsbuild, @cerminar, @franzoni, @Dr15Jones, @rcastello, @deguio, @slava77, @danduggan, @vadler, @mmusich, @davidlange6, @ktf, @nclopezo can you please review it and eventually sign? Thanks.
@ghellwig, @abbiendi, @Martin-Grunewald, @tlampen, @threus, @venturia, @frmeier, @battibass, @makortel, @acaudron, @dgulhan, @jlagram, @ferencek, @cerati, @trocino, @rociovilar, @kkrajczar, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @wmtford, @mmusich, @richard-cms, @imarches, @gbenelli, @appeltel, @rylanC24, @gpetruc, @matt-komm, @pvmulder, @bachtis this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: c7ef28a06c9d5de2e74faf921cbf6bd5ef578421
 I found an error when building: 
 
 <pre>Copying tmp/slc6_amd64_gcc491/src/SimMuon/MCTruth/src/SimMuonMCTruth/libSimMuonMCTruth.so to productstore area:
Leaving library rule at SimMuon/MCTruth
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/Validation/RecoTrack/src/MTVHistoProducerAlgoFactory.cc 
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/Validation/RecoTrack/src/MTVHistoProducerAlgoForTracker.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/Validation/RecoB/plugins/recoBSVTagInfoValidationAnalyzer.cc: In constructor 'recoBSVTagInfoValidationAnalyzer::recoBSVTagInfoValidationAnalyzer(const edm::ParameterSet&)':
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/Validation/RecoB/plugins/recoBSVTagInfoValidationAnalyzer.cc:95:121: error: no matching function for call to 'VertexClassifierByProxy<std::vector<reco::TemplatedSecondaryVertexTagInfo<reco::IPTagInfo<edm::RefVector<std::vector<reco::Track> >, reco::JTATagInfo>, reco::Vertex> > >::VertexClassifierByProxy(const edm::ParameterSet&)'
 recoBSVTagInfoValidationAnalyzer::recoBSVTagInfoValidationAnalyzer(const edm::ParameterSet& config) : classifier_(config)
                                                                                                                         ^
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/Validation/RecoB/plugins/recoBSVTagInfoValidationAnalyzer.cc:95:121: note: candidates are:
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/Validation/RecoB/plugins/recoBSVTagInfoValidationAnalyzer.cc:21:0:
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_4_X_2015-01-14-1400/src/SimTracker/TrackHistory/interface/VertexClassifierByProxy.h:20:5: note: VertexClassifierByProxy<Collection>::VertexClassifierByProxy(const edm::ParameterSet&, edm::ConsumesCollector&&) [with Collection = std::vector<reco::TemplatedSecondaryVertexTagInfo<reco::IPTagInfo<edm::RefVector<std::vector<reco::Track> >, reco::JTATagInfo>, reco::Vertex> >]
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7175/1872/summary.html

Dr15Jones committed:
VertexClassifierByProxy's constructor now needs a ConsumesCollector

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@davidlange6 I created a new SimDataFormats package for the new associators.

cmsbuild commented:
Pull request #7175 was updated. @civanch, @diguida, @lveldere, @cvuosalo, @ianna, @mdhildreth, @monttj, @cmsbuild, @cerminar, @franzoni, @Dr15Jones, @rcastello, @deguio, @slava77, @danduggan, @vadler, @mmusich, @davidlange6, @ktf, @nclopezo can you please check and sign again.

davidlange6 commented:
@Dr15Jones - impressive - thanks.

cmsbuild commented:
+1
Tested at: 594732563f79e3254557f4dc8e3cd9575ea49f30
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7175/1873/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7175/1873/summary.html

abbiendi commented:
Hi Chris, is there already a planned prerelease to integrate this ? since it affects all the track and muon validation although no changes are expected I was wondering if any special validation step is foreseen.

I am also working to improve something in the muon validation and it could be useful to me to understand the timing of the PRs. 

I have a few details to comment on SimMuon/MCTruth here below.
Thanks for your huge work.
   Giovanni

**** CSCHitAssociator.h/.cc ****

+ the new code lost a functionality which would be still useful for run1 scenario, related to the so-called "bad" CSC chambers, i.e. the ones existing in geometry but not providing hits in DIGI-RECO. 

see the current code:
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/interface/MuonTruth.h#L51
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonTruth.cc#L60 -L63

+ missing debug switch and related code  (bool flag "crossingframe" et al) which is instead present in DT and RPC associators. I would keep the current functionality as it might be useful for debug purposes.

https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/interface/MuonTruth.h#L72  - L76
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonTruth.cc#L13 -L16
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonTruth.cc#L27 -L39
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonTruth.cc#L67 -L95

**** MuonAssociatorByHitsHelper.h/.cc   <-> MuonAssociatorByHits.h/.cc  ****

there are some debug switches and parameters that are lost:

+ debug switch "dumpInputCollections" and parameters "crossingframe" and others are lost in the migration. I would keep them as in the current code:
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/interface/MuonAssociatorByHits.h#L110
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L37  -L40
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L63 -L70
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L100 -L103
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L201 -L282

+ setting of internal variable printRtS should be true in:
https://github.com/Dr15Jones/cmssw/blob/eventBasedTrackAssociation/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L66
as in:
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L193

(the line : https://github.com/Dr15Jones/cmssw/blob/eventBasedTrackAssociation/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L59
could be dropped as it the variable is not used before L66)

ianna commented:
+1

Dr15Jones commented on SimMuon/MCTruth/src/CSCHitAssociator.cc:
```diff
@@ -0,0 +1,108 @@
+  // get CSC Geometry to use CSCLayer methods
+  edm::ESHandle<CSCGeometry> mugeom;
+  setup.get<MuonGeometryRecord>().get( mugeom );
+  cscgeom = &*mugeom;
```
>the new code lost a functionality which would be still useful for run1 scenario, related to the so-called "bad" CSC chambers, i.e. the ones existing in geometry but not providing hits in DIGI-RECO.
see the current code:
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/interface/MuonTruth.h#L51
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonTruth.cc#L60 -L63

Dr15Jones replied:
The variable ```MuonTruth::cscBadChambers``` was set but never read by ``MuonTruth```.  ```CSCHitAssociator``` just contains exact copies of ```MuonTruth::associateCSCHitId``` and ```MuonTruth::associateHitId``` with only those same member data from ```MuonTruth``` needed by those routines.

Dr15Jones replied:
```MuonTruth``` still maintains the variable ```cscBadChambers``` and it is used by ```MuonAssociatorByHitsHelper``` exactly the way it was previously used in ```MuonAssociatorByHits```.

Dr15Jones replied:
I see now that ```MuonTruth::cscBadChambers``` was only called from code that is commented out in both ```MuonAssociatorByHits``` and ```MuonAssociatorByHitsHelper``` so was never actually used.

Dr15Jones commented on SimMuon/MCTruth/src/MuonAssociatorByHitsHelper.cc:
```diff
@@ -0,0 +1,1154 @@
+    EfficiencyCut_muon = 0.;
+  }
+
+}
```
>there are some debug switches and parameters that are lost:
>
>debug switch "dumpInputCollections" and parameters "crossingframe" and others are lost in the migration. I would keep them as in the current code:
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/interface/MuonAssociatorByHits.h#L110
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L37 -L40
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L63 -L70
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L100 -L103
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/SimMuon/MCTruth/src/MuonAssociatorByHits.cc#L201 -L282

Dr15Jones replied:
Looks like ```dumpInputCollections``` can be added back without causing any knock on effects.

Dr15Jones replied:
I was wrong on the above. It is not possible to add ```crossingframe``` since that would require getting more data from the Event. As ```crossingframe``` is only used if ```dumpInputCollections``` is true, would ```dumpInputCollections``` still be useful without ```crossingframe```?

Dr15Jones replied:
Similarly, the last two remaining parameters ```simtrackstag``` and ```simtracksXFTag``` are only used if ```crossingframe``` is true and are only used to dump debug information.

Dr15Jones replied:
My suggestion is it would be better to make an ```EDAnalyzer``` which dumps this information given the same InputTags.

Dr15Jones replied:
I added back the ability to get the diagnostic print outs. It is now done by passing an optional std::function to MuonAssociatorByHitsHelper which can do the dump.

Dr15Jones commented on SimMuon/MCTruth/src/MuonAssociatorByHits.cc:
```diff
@@ -137,1349 +46,77 @@ MuonAssociatorByHits::associateRecoToSim( const edm::RefToBaseVector<reco::Track
-  MapOfMatchedIds tracker_matchedIds_INVALID, muon_matchedIds_INVALID;
-
-  IndexAssociation     outputCollection;
   bool printRtS(true);
```
> this line could be dropped as it the variable is not used before L66

Dr15Jones commented on SimMuon/MCTruth/src/MuonAssociatorByHits.cc:
```diff
@@ -137,1349 +46,77 @@ MuonAssociatorByHits::associateRecoToSim( const edm::RefToBaseVector<reco::Track
   // DT hit association
-  printRtS = true;
-  DTHitAssociator dttruth(*e,*setup,conf_,printRtS);
+  printRtS = false;
```
> setting of internal variable printRtS should be true

Dr15Jones replied:
OK

Dr15Jones commented on SimMuon/MCTruth/src/MuonAssociatorByHits.cc:
```diff
@@ -137,1349 +46,77 @@ MuonAssociatorByHits::associateRecoToSim( const edm::RefToBaseVector<reco::Track
-	    
-	    // discard BAD CSC chambers (ME4/2) from hit counting
-	    if (dId.subdetId() == MuonSubdetId::CSC) {
-              if (csctruth.cscBadChambers->isInBadChamber(CSCDetId(dId))) {
```
This is the only place ```cscBadChambers``` was ever accessed and it is commented out.

abbiendi replied:
Hi Chris,

sorry I had forgot...you are right. The functionality is lost from the time of the TrackingParticle redesign about one year a half ago.
Recovering it would require some non-trivial code that I do not think it is worthwhile right now.
cheers

 Giovanni


On Jan 16, 2015, at 9:11 PM, Chris Jones wrote:


In SimMuon/MCTruth/src/MuonAssociatorByHits.cc<https://github.com/cms-sw/cmssw/pull/7175#discussion-diff-23106047>:

> -               newhit = false;
> -           //grouped, splitting
> -           if (UseGrouped && UseSplitting)
> -             newhit = true;
> -         }
> -         if (newhit) {
> -           tphits.push_back(*TPhit);
> -         }
> -         delete stripDetId;
> -       }
> -       else if (detector == DetId::Muon) {
> -         n_muon_simhits++;
> -
> -         // discard BAD CSC chambers (ME4/2) from hit counting
> -         if (dId.subdetId() == MuonSubdetId::CSC) {
> -              if (csctruth.cscBadChambers->isInBadChamber(CSCDetId(dId))) {


This is the only place cscBadChambers was ever accessed and it is commented out.

—
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/7175/files#r23106047>.

Dr15Jones committed:
Print diagnostic messages when problems occur in MuonAssociation

slava77 commented:
I suppose more changes are coming, so no need to sign off yet.

Chris, this is one of the PRs where you should also post a (random text ok) message on github after a commit to make cms-bot notice changes.

As of https://github.com/cms-sw/cmssw/pull/7175	5947325
I see in 25202 on 100 events that the memory increased by ~1%, could be a fluc though.
- Max VSIZ 3447.7 on evt 52 ; max RSS 2923.22 on evt 87
- changes to 
- Max VSIZ 3488.67 on evt 88 ; max RSS 2956.66 on evt 87

Based on our extended stat tests (https://github.com/slava77/cms-reco-tools/blob/master/extendedTests.sh)
there are no changes in the monitored values in DQM or with FWlite, except for the expected:
- there is a change in log warnings, changing from ```QuickTrackAssociatorByHits: MultiTrackValidator:hltPixelTracksV@streamBeginRun``` to ```QuickTrackAssociatorByHitsImpl: QuickTrackAssociatorByHitsProducer:hltTrackAssociatorByHits@ctor```

cmsbuild commented:
Pull request #7175 was updated. @civanch, @diguida, @lveldere, @cvuosalo, @ianna, @mdhildreth, @monttj, @cmsbuild, @cerminar, @franzoni, @Dr15Jones, @rcastello, @deguio, @slava77, @danduggan, @vadler, @mmusich, @davidlange6, @ktf, @nclopezo can you please check and sign again.

Dr15Jones committed:
Added back the ability to print diagnostics for MuonAssociatorByHits

The MuonAssociatorByHitsHelper now takes a std::function which if set
is called when an association is calculated. This allows users of the
helper to pass in a functor which prints the diagnostic output which
was in the previously MuonAssociatorByHits implementation.
The parameter used to turn on the diagnostics, dumpInputCollections,
was changed to be untracked.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #7175 was updated. @civanch, @diguida, @lveldere, @cvuosalo, @ianna, @mdhildreth, @monttj, @cmsbuild, @cerminar, @franzoni, @Dr15Jones, @rcastello, @deguio, @slava77, @danduggan, @vadler, @mmusich, @davidlange6, @ktf, @nclopezo can you please check and sign again.

Dr15Jones commented:
I'm now done with the changes from the comments given.

I added back the ability to turn on the diagnostics. This now makes use of helper classes to do the actual dumping if needed.

I do not plan on added back getting the CSCBadChambers since after they were retrieved no code in CMSSW ever used it. If that functionality is needed in the future then the code which wants that information should retrieved it from the EventSetup. That is my recommendation as the framework L2.

Dr15Jones commented:
+1
for geometry so Yana doesn't have to do it again :)

cmsbuild commented:
+1
Tested at: d27695186f631bde412f573769b2d625b3236d95
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7175/1972/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7175/1972/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/7175	d276951
tested in CMSSW_7_4_0_pre3 /test area sign487/

results are the same as in the summary above https://github.com/cms-sw/cmssw/pull/7175#issuecomment-70371409

Dr15Jones commented:
ping @civanch, @diguida, @lveldere, @mdhildreth, @monttj, @cerminar, @franzoni, @rcastello, @deguio, @danduggan, @vadler, @mmusich

deguio commented:
+1

lveldere commented:
+1

civanch commented:
+1

davidlange6 commented:
bypassing the remaining signatures on this one given the lack of complaints (so that it doesn't get out of sync and create more work)

ktf commented:
Apparently this one breaks one test in the gcc481 build due to some unused variable.

@Dr15Jones can you have a look?

Dr15Jones commented:
gcc4.8.1 compile problem fixed in #7340

kkrajczar commented:
Dear all,

It seems that this PR broke the HI RelVals. The related error message
that we see is "RuntimeError: An entry in sequence tracksValidationFS
has no label". Please find below a description of the problem and a
pointer to a PR with a workaround.

In Thursday's IB (CMSSW_7_4_X_2015-01-22-0200) the following cmsDriver
command correctly produces a python config:

cmsDriver.py step3 --conditions auto:run2_mc_HIon -s
RAW2DIGI,L1Reco,RECO,VALIDATION,DQM -n 2 --eventcontent RECOSIM,DQM
--scenario HeavyIons --datatier GEN-SIM-RECO,DQMIO --customise
SLHCUpgradeSimulations/Configuration/postLS1Customs.customisePostLS1
--magField 38T_PostLS1 --no_exec

However, the same command in Friday's IB (CMSSW_7_4_X_2015-01-23-0200)
crashes with the error message "RuntimeError: An entry in sequence
tracksValidationFS has no label". It seems that the crash is caused by
this PR, as Thursday's build + PR #7175 results in the
above-mentioned error, while Thursday's build itself is ok.

We don't really understand the reason of the error, but we prepared a
workaround PR #7404 for the HI Validation setup, so that the issue does not
shadow further problems with upcoming PRs. We offer our help for testing
in any level, however we really need expertise help in the actual fix.

(@yetkinyilmaz and @BetterWang probably want to follow this)

makortel commented:
While investigating something unrelated within the track-TP associations, I noticed that we have still the old ESProduct-based implementations in `SimTracker/TrackAssociation` (that are, so far, nearly exact copies of these ones). Are the old ones still going to be supported/(used), or could they be cleaned up? I'm just afraid that the current situation will eventually lead to confusion (especially if/when something needs to be changed in them).
